File: app/src/main/java/no/nav/tilbakemeldingsmottak/consumer/pdl/PdlService.java
package no.nav.tilbakemeldingsmottak.consumer.pdl;

import com.graphql_java_generator.exception.GraphQLRequestExecutionException;
import com.graphql_java_generator.exception.GraphQLRequestPreparationException;
import lombok.RequiredArgsConstructor;
import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorException;
import no.nav.tilbakemeldingsmottak.exceptions.ErrorCode;
import no.nav.tilbakemeldingsmottak.graphql.IdentGruppe;
import no.nav.tilbakemeldingsmottak.graphql.IdentInformasjon;
import no.nav.tilbakemeldingsmottak.graphql.Identliste;
import no.nav.tilbakemeldingsmottak.graphql.util.QueryExecutor;
import org.springframework.stereotype.Service;

import java.util.List;

// Relevante linker:
// https://github.com/graphql-java-generator/graphql-maven-plugin-project/wiki/client_spring#the-spring-conf
// https://github.com/graphql-java-generator/GraphQL-Forum-Maven-Tutorial-client
@Service
@RequiredArgsConstructor
public class PdlService {
    private final QueryExecutor queryExecutor;

    public Identliste hentIdenter(String ident, List<IdentGruppe> grupper) throws ClientErrorException {
        try {
            return queryExecutor.hentIdenter("{identer {ident gruppe historisk}}", ident, grupper, false);
        } catch (GraphQLRequestExecutionException | GraphQLRequestPreparationException e) {
            throw new ClientErrorException("Graphql query mot PDL feilet", e, ErrorCode.PDL_ERROR);
        } catch (Exception ex) {
            throw new ClientErrorException("Ukjent feil mot PDL", ex, ErrorCode.PDL_ERROR);
        }
    }

    public String hentAktorIdForIdent(String ident) throws ClientErrorException {
        List<IdentInformasjon> identer = hentIdenter(ident, List.of(IdentGruppe.AKTORID)).getIdenter();

        if (identer == null || identer.isEmpty()) {
            throw new ClientErrorException("Fant ingen aktørId for ident", ErrorCode.PDL_MISSING_AKTORID);
        }

        return identer.get(0).getIdent();
    }

}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/consumer/pdl/PdlService.kt
package no.nav.tilbakemeldingsmottak.consumer.pdl

import no.nav.tilbakemeldingsmottak.consumer.pdl.domain.IdentDto
import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorException
import no.nav.tilbakemeldingsmottak.exceptions.ErrorCode
import no.nav.tilbakemeldingsmottak.metrics.MetricLabels
import no.nav.tilbakemeldingsmottak.metrics.Metrics
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service

@Service
class PdlService(private val pdlClient: PdlClient) {

    private val log = LoggerFactory.getLogger(javaClass)

    @Value("\${pdl.url}")
    lateinit var pdlUrl: String

    @Metrics(
        value = MetricLabels.DOK_CONSUMER,
        extraTags = [MetricLabels.PROCESS_CODE, "hentAktoerIdForIdent"],
        percentiles = [0.5, 0.95],
        histogram = true
    )
    @Cacheable("hentIdenter")
    fun hentPersonIdents(brukerId: String): List<IdentDto> {
        return pdlClient.performQuery(brukerId)?.hentIdenter?.identer?.map {
            IdentDto(
                it.ident,
                it.gruppe.toString(),
                it.historisk
            )
        } ?: listOf(IdentDto(brukerId, "AKTORID", false))
    }


    fun hentAktorIdForIdent(ident: String): String {
        log.info("Skal hente aktørId for ident")
        val identer = hentPersonIdents(ident)
        if (identer.isEmpty()) {
            throw ClientErrorException("Fant ingen aktørId for ident", null, ErrorCode.PDL_MISSING_AKTORID)
        }
        return identer[0].ident
    }
}
